services/dynamic/dynamic-toml.tester.js

```diff
@@ -0,0 +1,107 @@
+import { createServiceTester } from '../tester.js'
+export const t = await createServiceTester()
+
+t.create('No URL specified')
+  .get('.json?query=$.name&label=Package Name')
+  .expectBadge({
+    label: 'Package Name',
+    message: 'invalid query parameter: url',
+    color: 'red',
+  })
+
+t.create('No query specified')
+  .get(
+    '.json?url=https://raw.githubusercontent.com/squirrelchat/smol-toml/mistress/bench/testfiles/toml-spec-example.toml&label=Package Name',
```

Comment:
I'm not sure I understand...
All these tests use the json endpoint?

Reply:
The yml and xml do it too, I don't really know why

Reply:
Just to explain what is going on here:

You can call any shields badge URL with a `.json` extension to request a JSON response instead of SVG. So for example https://img.shields.io/github/v/tag/badges/shields (SVG badge) https://img.shields.io/github/v/tag/badges/shields.json (same thing as JSON)

In the service tests we use this because it is easier to make assertions about the JSON response than the SVG response. The logic for rendering that object into a SVG is tested separately.

`createServiceTester()` makes a test object with a base of `/badge/dynamic/toml` so every url we `.get()` is relative to `/badge/dynamic/toml`

The dynamic badges don't take any positional parameters - only query string parameters.

So when we call 

```
.get(
  '.json?url=https://raw.githubusercontent.com/squirrelchat/smol-toml/mistress/bench/testfiles/toml-spec-example.toml&label=Package Name',
)
```

we're really requesting

`/badge/dynamic/toml.json?url=https%3A//raw.githubusercontent.com/squirrelchat/smol-toml/mistress/bench/testfiles/toml-spec-example.toml&label=Package%20Name`

which returns a response like

https://img.shields.io/badge/dynamic/toml.json?url=https%3A//raw.githubusercontent.com/squirrelchat/smol-toml/mistress/bench/testfiles/toml-spec-example.toml&label=Package%20Name

..and then we're making assertions about it.

Reply:
Thanks!